### PR TITLE
Autogenerate the SDK from the OpenAPI spec

### PR DESCRIPTION
**Pull request by AndrewBenzSW**
This PR replaces the existing SDK with a new one that's been autogenerated from the ShipEngine OpenAPI spec.  The existing methods and models are still here, but have been marked as Obsolete in favor of the new methods.  The autogeneration is using custom modified [templates](https://openapi-generator.tech/docs/templating/) for the [CSharp](https://openapi-generator.tech/docs/generators/csharp/) [openapi-generator](https://openapi-generator.tech/) tool.  The major benefit of this is that we don't have to maintain as much code and we get an SDK for the entire API as opposed to just a handful of methods.

This SDK will now be only as good as our documentation as it's based off of it.  I've already found an instance where what's documented is not correct: when you list labels, our documentation says the ids of the packages will be `se-*****` strings, but in reality, they are numbers.  Since the existing API methods are still here, I don't think that should hold us up from moving forward with this.

The underlying HTTP mechanism is remaining the same.  I wrote the API method templates to make use of our existing client code, so all the serialization, http, and retry logic is being leveraged.

One note: the docs, api methods, and models will get regenerated every time we make an update, but tests will not.  So manual changes to tests are encouraged as they will stick around.

### CODE REVIEW

File: ShipEngineSDK.Test/Api/AccountApiTests.cs [hunk 1]
```diff
@@ -0,0 +1,127 @@
+/*
```
Comment (joshuaflanagan) [inline ShipEngineSDK.Test/Api/AccountApiTests.cs:1]: We should not commit auto-generated tests - especially ones that don't do anything.

File: ShipEngineSDK/Api/CarriersApi.cs [hunk 2]
```diff
@@ -0,0 +1,631 @@
...
+    /// <summary>
+    /// Add Funds To Carrier Add Funds To A Carrier
+    /// </summary>
+    /// <exception cref="System.ArgumentNullException">Thrown when required argument is null</exception>
+    /// <exception cref="ShipEngineSDK.ShipEngineException">Thrown when fails to make API call</exception>
+    /// <param name="carrierId">Carrier ID</param>
+    /// <param name="addFundsToCarrierRequestBody"></param>
+    /// <param name="cancellationToken">Cancellation Token to cancel the request.</param>
+    /// <returns>Task of ApiResponse (AddFundsToCarrierResponseBody)</returns>
+    Task<AddFundsToCarrierResponseBody> AddFundsToCarrier(Config methodConfig, string carrierId, AddFundsToCarrierRequestBody addFundsToCarrierRequestBody, CancellationToken cancellationToken = default);
```
Comment (joshuaflanagan) [inline ShipEngineSDK/Api/CarriersApi.cs:49]: If we are breaking compatibility anyway, I think we should get rid of the ability to pass a `Config` to methods.  It has a very limited arbitrary set of things it lets you change. There are better ways to modify the way the request is performed (like passing an `Action<HttpRequestMessage>` or similar.

File: generation/templates/auth/OAuthAuthenticator.mustache [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+            return new HeaderParameter(KnownHeaders.Authorization, token);
+        }
+
+        /// <summary>
+        /// Gets the token from the OAuth2 server.
+        /// </summary>
+        /// <returns>An authentication token.</returns>
+        async Task<string> GetToken()
+        {
+            var client = new RestClient(_tokenUrl,
```
Comment (joshuaflanagan) [inline generation/templates/auth/OAuthAuthenticator.mustache:76]: Why is this file included? How is the OAuthAuthenticator used with ShipEngine? Why does it use RestClient instead of HttpClient?

File: generation/templates/git_push.sh.mustache [hunk 4]
```diff
@@ -0,0 +1,57 @@
+#!/bin/sh
```
Comment (joshuaflanagan) [inline generation/templates/git_push.sh.mustache:1]: This seems unnecessary.

Comment (joshuaflanagan) [reply generation/templates/git_push.sh.mustache:1]: I'm _pretty sure_ (worth testing) you can delete any templates that you did not modify, and it will fallback to the built-in ones. It would be much easier to manage if we have a smaller subset of files we know we need to take care of.
For files that we do modify, it would be useful to commit the original unmodified template, and then commit our changes, so it easy to see our diff. That will make it easier to update templates in the future (if they are changed upstream).

File: docs/AccountApi.md [hunk 5]
```diff
@@ -0,0 +1,431 @@
+# ShipEngineSDK.Api.AccountApi
```
Comment (joshuaflanagan) [inline docs/AccountApi.md:1]: Are these docs intendede to be published somewhere, or are they for browsing on github?

We should make it very clear they are auto-generated and not intended to be modified. Maybe using [front matter](https://jekyllrb.com/docs/front-matter/)?

**Comment by joshuaflanagan** [comment]
We should remove the obsolete warnings (use a pragma/directive, or some other mechanism of disabling it globally) for the methods we marked obsolete. It leads to misleading/noisy build output.

File: generation/templates/git_push.sh.mustache (hunk 4 above)
Comment (AndrewBenzSW) [reply generation/templates/git_push.sh.mustache:1]: I like that.  I'll create a new PR with just the templates that will get modified since we squash individual commits.

File: ShipEngineSDK/Api/CarriersApi.cs (hunk 2 above)
Comment (AndrewBenzSW) [reply ShipEngineSDK/Api/CarriersApi.cs:49]: I'm happy to get rid of this overload.  I just added it to be as close to how the API currently works.

File: generation/templates/auth/OAuthAuthenticator.mustache (hunk 3 above)
Comment (AndrewBenzSW) [reply generation/templates/auth/OAuthAuthenticator.mustache:76]: It's not necessary, but it's part of the default set of templates.  After #98 gets approved, I'm going to remove this and all other unmodified templates.

File: docs/AccountApi.md (hunk 5 above)
Comment (AndrewBenzSW) [reply docs/AccountApi.md:1]: They're just meant for browsing on Github (or in your local repo).  Front matter looks like it'll work on GH but I'll see what it looks like rendering locally.  I also came across a type of Markdown comment that I'll try as well to see what works & looks better.

**Comment by joshuaflanagan** [comment]
I'm not sure it's worth keeping them, if we don't have a clear need. They
don't add much more than the in code doc comments/intellisense, and can
potentially compete with or cause confusion with the actual ShipEngine API
documentation site.

File: ShipEngineSDK/Api/CarriersApi.cs (hunk 2 above)
Comment (joshuaflanagan) [reply ShipEngineSDK/Api/CarriersApi.cs:49]: RIght. It is something I wanted to get rid of, but would mean a big breaking change. But if we are making a breaking change anyway, we should strongly consider if it is the best way to convey that type of information. 
@jpill may have thoughts on the subject.

**Comment by AndrewBenzSW** [comment]
> I'm not sure it's worth keeping them, if we don't have a clear need. They don't add much more than the in code doc comments/intellisense, and can potentially compete with or cause confusion with the actual ShipEngine API documentation site.
> […](#)

I think the benefit is that some developers may prefer SDK specific docs in the Github repo vs. the other methods mentioned.  It can be more straightforward to look at a document to see what's in a class instead of going through Intellisense entries one by one or trawling through comments on the actual source.  And while it shouldn't be difficult to figure out how to use the SDK by looking at the API docs, it's extra friction that doesn't seem necessary.  I was able to copy & paste an example from the generated docs and after replacing my api key, it just worked.  That's certainly more convenient than looking at the APIs online, figuring out how it maps to the SDK, and then trying to make that work.

I know they're not perfect.  I've already seen some instances where strings aren't quoted and enum values aren't assigned correctly.  But those things are easy to fix if we come across them and since this is a public repo, others can help us fix them.

File: ShipEngineSDK/Api/CarriersApi.cs (hunk 2 above)
Comment (jpill) [reply ShipEngineSDK/Api/CarriersApi.cs:49]: I'm not married to keeping the `Config` approaches, I agree that there are now improved paths to accomplish what that does.

**Comment by AndrewBenzSW** [comment]
I've removed the `[Obsolete]` tags for now and just bumped the version to `2.1.0`.   This will let us use and iterate on the new SDK methods while not nudging anyone else to move yet.

**Review by joshuaflanagan — APPROVED** [review]
